core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/marshall/QueryParamMarshaller.java

```diff
@@ -60,6 +61,13 @@ public final class QueryParamMarshaller {
         }
     };
 
+    public static final JsonMarshaller<Void> NULL = (val, context, paramName, sdkField) -> {
+        if (context.protocolHandler().isTraitValidationEnabled(RequiredTrait.class)
+            && sdkField.containsTrait(RequiredTrait.class)) {
```

Comment:
Hmm, can we simplify this and only attach the RequiredTrait to protocols for which we have it enabled? 

That way we don't need an explicit isTraitValidationEnabled check at runtime. If the RequiredTrait is there, we do it. Otherwise, we don't.

Reply:
Made change with a separate commit [dbc46a9](https://github.com/aws/aws-sdk-java-v2/pull/3546/commits/dbc46a9cb22bfb034f95adbd1c980fc15f74281f) for easier diffing.